### ISSUE TITLE
Compile error with mingw64 / wsl (OCD-1042)

### ISSUE THREAD

**Issue by ddemelier-ts** [espressif/openocd-esp32#347]
### Development Kit

esp32c6-devkitc

### Module or chip used

esp32c6

### Debug Adapter

ftdi

### OpenOCD version

git

### Operating System

Windows

### Using an IDE ?

no

### OpenOCD command line

no

### JTAG Clock Speed

no

### ESP-IDF version

no

### Problem Description

Cross-compiling openocd-esp32 under wsl with x86_64-w64-mingw32 fails with the following error:

```
libtool: compile:  x86_64-w64-mingw32-gcc -DHAVE_CONFIG_H -I. -D__USE_MINGW_ANSI_STDIO -DFD_SETSIZE=128 -I./src -I./src -DPKGDATADIR=\"/usr/share/openocd\" -DBINDIR=\"/usr/bin\" -I./jimtcl -I./jimtcl -Wall -Wstrict-prototypes -Wformat-security -Wshadow -Wextra -Wno-unused-parameter -Wbad-function-cast -Wcast-align -Wredundant-decls -Wpointer-arith -Wundef -Werror -g -O2 -MT src/target/espressif/esp_semihosting.lo -MD -MP -MF src/target/espressif/.deps/esp_semihosting.Tpo -c src/target/espressif/esp_semihosting.c -o src/target/espressif/esp_semihosting.o
src/target/espressif/esp_semihosting.c: In function ‘esp_semihosting_common’:
src/target/espressif/esp_semihosting.c:25:36: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
   25 | #define fsync(fd) FlushFileBuffers((HANDLE)fd)
      |                                    ^
src/target/espressif/esp_semihosting.c:730:26: note: in expansion of macro ‘fsync’
  730 |    semihosting->result = fsync(fd);
      |                          ^~~~~
cc1: all warnings being treated as errors
```

### Expected behavior

Doesn't build.

**Comment by erhankur** [comment]
I can not try on windows now but can suggest you 2 solutions

1- Modify the `fsync(fd)` definition as below. esp_semihosting.c:25
```
#define fsync(fd) FlushFileBuffers((HANDLE)(uintptr_t)(fd))
```
2 - Reconfigure the project by disabling the compiler error
```
./configure CFLAGS="-Wno-int-to-pointer-cast"
```
Let me know if these solutions work for you.

**Comment by ddemelier-ts** [comment]
No I think the issue comes from the fact that `HANDLE` is a `void *` while `semihosting_get_field` returns `uint64_t`. This function should return a generic `fd` type but that's a bigger work. Would this be correct?

```diff
--- src/target/espressif/esp_semihosting.c.orig 2024-11-22 16:39:58.688979902 +0100
+++ src/target/espressif/esp_semihosting.c      2024-11-22 16:41:50.689353023 +0100
@@ -726,10 +726,10 @@
                }
                retval = semihosting_read_fields(target, 1, fields);
                if (retval == ERROR_OK) {
-                       int fd = semihosting_get_field(target, 0, fields);
-                       semihosting->result = FlushFileBuffers((uintptr_t)fd);
+                       uint64_t fd = semihosting_get_field(target, 0, fields);
+                       semihosting->result = fsync(fd);
                        semihosting->sys_errno = errno;
-                       LOG_DEBUG("fsync('%d')=%" PRId64, fd, semihosting->result);
+                       LOG_DEBUG("fsync('%'" PRIu64 "')=%" PRId64, fd, semihosting->result);
                }
                break;
```

**Comment by erhankur** [comment]
Yes. Your solution is correct. We will fix the fd type there.

Regarding Zephyr, it is not fully supported yet, but we do have an initial code base. We plan to improve it early next year.

**Comment by erhankur** [comment]
Fixed id a13942993803ca90eda48a1b0a47d3df32a1277e
